src/socket.c

```diff
@@ -731,26 +719,43 @@ void pthreads_socket_select(zval *read, zval *write, zval *except, uint32_t sec,
 	}
 
 	PHP_SAFE_MAX_FD(mfd, 0);
-	
-	if (sec || usec) {
+
+	/* If seconds is not set to null, build the timeval, else we wait indefinitely */
+	if (sec != NULL) {
+		zval tmp;
+
+		if (Z_TYPE_P(sec) != IS_LONG) {
+			tmp = *sec;
+			zval_copy_ctor(&tmp);
+			convert_to_long(&tmp);
+			sec = &tmp;
+		}
+
+		/* Solaris + BSD do not like microsecond values which are >= 1 sec */
 		if (usec > 999999) {
-			tv.tv_sec = sec + (usec / 1000000);
+			tv.tv_sec = Z_LVAL_P(sec) + (usec / 1000000);
 			tv.tv_usec = usec % 1000000;
 		} else {
-			tv.tv_sec = sec;
+			tv.tv_sec = Z_LVAL_P(sec);
 			tv.tv_usec = usec;
 		}
+
+		tv_p = &tv;
+
+		if (sec == &tmp) {
+			zval_dtor(&tmp);
```

Comment:
Longs are not reference counted, and so there's no need to destroy them (the same operation in php-src looks superfluous - probably outdated code from PHP 5, when every type was reference counted).

Reply:
Good to know. I've removed the `zval_dtor` call.